docs/typescript.md

```diff
@@ -1,14 +1,597 @@
+// 'getNftInformation' is a private method of class 'NftController'
+sinon.stub(nftController, 'getNftInformation');
+```
+
+🚫 `as any`
+
+```typescript
+sinon.stub(nftController, 'getNftInformation' as any);
+```
+
+✅ Compiler specifies that the target type should be `keyof NftController`
+
+```typescript
+sinon.stub(nftController, 'getNftInformation' as keyof typeof nftController);
+```
+
+- Even an assertion to a wrong type still allows the compiler to show us warnings and errors as the code changes, and is therefore preferrable to the dangerous radio silence enforced by `any`.
+- For type assertions to an incompatible shape, use `as unknown as` as a last resort rather than `any` or `as any`.
+
+##### `as` is always acceptable to use for TypeScript syntax other than type assertion
+
+- Writing type guards often requires using the `as` keyword.
+
+```typescript
+function isFish(pet: Fish | Bird): pet is Fish {
+  return (pet as Fish).swim !== undefined;
+}
+```
+
+- Key remapping in mapped types uses the `as` keyword.
+
+```typescript
+type MappedTypeWithNewProperties<Type> = {
+  [Properties in keyof Type as NewKeyType]: Type[Properties];
+};
+```
+
+##### `as` is acceptable to use for typing data objects whose shape and contents are determined at runtime, externally, or through deserialization
+
+Preferably, this typing should be accompanied by runtime schema validation performed with type guards and unit tests.
+
+- e.g. The output of `JSON.parse()` or `await response.json()` for a known JSON input.
+- e.g. The type of a JSON file.
+
+<!-- TODO: Add example -->
+
+##### `as` may be acceptable to use in tests, for mocking or to exclude irrelevant but required properties from an input object
+
+It's recommended to provide accurate typing if there's any chance that omitting properties affects the accuracy of the test.
+
+If that is not the case, however, mocking only the properties needed in the test improves readability. It makes the intention and scope of the mock clear, and is more convenient to write, which can encourage test writing and improve test coverage.
+
+<!-- TODO: Add examples -->
+
+### Compiler Directives
+
+TypeScript provides several directive comments that can be used to suppress TypeScript compiler errors. Using these to ignore typing issues is dangerous and reduces the overall effectiveness of TypeScript.
+
+#### Acceptable usages of `@ts-expect-error`
+
+##### Use `@ts-expect-error` to force runtime execution of a branch for validation or testing
+
+Sometimes, there is a need to force a branch to execute at runtime for security or testing purposes, when that branch has correctly been inferred as being inaccessible by the TypeScript compiler.
+
+Suppressing compiler errors to avoid typing issues is usually dangerous because it results in a loss of information and safety. However, consciously using `@ts-expect-error` to add runtime checks represents an improvement in these aspects, and therefore is not discouraged.
```

Comment:
Nit: In general it's fair to say that suppressing compiler errors "results in a loss of information and safety", but I don't think that applies to unreachable branches. I don't think that's adequately explained here ("prepresents an improvement in these aspects" is pretty vague), but I can't think of a great way to explain it concisely either, and it doesn't seem totally necessary. So maybe we can drop this paragraph instead:

```suggestion
```